### PR TITLE
Userscripts, suggest command, winid env

### PR DESCRIPTION
**Pull request by seebye**
The following pull request contains the following changes:

- userscripts are called with an additional environment variable "QUTE_WID" = x11 window id
  (e.g. allows mpv to draw on qutebrowsers window)
- there's a new command suggest which allows userscripts to create a menu in qutebrowser
  (examples: see userscript iframe, vid)
- userscript iframe: displays every iframe url and opens the choosen one
- userscript vid: tries to play a video found in the source of the site or let's the user choose one
  * it also allows to choose a specific behavior e.g. looping the file or keeping mpv open
  * it can use qutebrowsers window to display the video
  * it has a picture in picture mode (you likely need to create the right rules in e.g. bspwm, i3wm, etc.)
  * it uses hooktube to surpass country based takedowns (youtube)

---
This change is [Reviewable](https://reviewable.io/reviews/qutebrowser/qutebrowser/3376)

### CODE REVIEW

File: qutebrowser/browser/commands.py [hunk 1]
```diff
@@ -1231,6 +1250,12 @@ def _run_userscript(self, cmd, *args, verbose=False):
             'QUTE_MODE': 'command',
         }
 
+        try:
+            # not sure if int(winId()) is safe to call on every os.
+            env['QUTE_WID'] = str(int(self._tabbed_browser.winId()))
```
Comment (rcorre) [inline qutebrowser/browser/commands.py]: `CommandDispatcher` has a `_win_id` attribute.

File: qutebrowser/completion/models/miscmodels.py [hunk 2]
```diff
@@ -47,6 +47,31 @@ def helptopic(*, info):
...
+        model = completionmodel.CompletionModel(
+            column_widths=tuple(col_size for i in range(cols)))
+        raw = (args[4][1:-1]
+               if any(args[4].startswith(i) for i in ('"', "'"))
+               else args[4])
+        suggestions = (i.split(sep_col) for i in raw.split(sep_sug))
+
+        model.add_category(listcategory.ListCategory(
+            "Suggestions", suggestions))
+    except Exception as e:
```
Comment (rcorre) [inline qutebrowser/completion/models/miscmodels.py]: We should be more specific about the exception/message here. For example, if the arg count isn't right, then specifically catch and report that.

File: qutebrowser/browser/commands.py [hunk 3]
```diff
@@ -1175,6 +1175,25 @@ def tab_move(self, index: typing.Union[str, int] = None, count=None):
...
+    def suggest(self, command: str, columns: int,
+                sep_suggestions: str, sep_columns: str,
+                suggestions: str, choice: str):
+        """Allows userscripts to let a user choose between extracted information.
+
+        Args:
+            command: Qutebrowser command to run with the choice
+            columns: self-explanatory
+            sep_suggestions: charsequence which separates suggestions
+            sep_columns: charsequence which separates columns
```
Comment (rcorre) [inline qutebrowser/browser/commands.py]: I'd just call it a `string`, not a `charsequence`.

File: qutebrowser/browser/commands.py [hunk 4]
```diff
@@ -1175,6 +1175,25 @@ def tab_move(self, index: typing.Union[str, int] = None, count=None):
...
+    @cmdutils.register(instance='command-dispatcher', scope='window')
+    @cmdutils.argument('choice', completion=miscmodels.suggest)
+    def suggest(self, command: str, columns: int,
+                sep_suggestions: str, sep_columns: str,
+                suggestions: str, choice: str):
+        """Allows userscripts to let a user choose between extracted information.
+
+        Args:
+            command: Qutebrowser command to run with the choice
+            columns: self-explanatory
```
Comment (rcorre) [inline qutebrowser/browser/commands.py]: Can't you just count the number of column separators in the first row?

File: qutebrowser/browser/commands.py [hunk 5]
```diff
@@ -1175,6 +1175,25 @@ def tab_move(self, index: typing.Union[str, int] = None, count=None):
...
+            sep_suggestions: charsequence which separates suggestions
```
Comment (rcorre) [inline qutebrowser/browser/commands.py]: maybe these could be flags with a reasonable default (`sep_suggestions=';',sep_columns=',')?
It might be even simpler to use a single separator character and double it to represent a new row.
e.g. `foo,bar,baz,,biz,buz,baa`

File: qutebrowser/browser/commands.py [hunk 6]
```diff
@@ -1175,6 +1175,25 @@ def tab_move(self, index: typing.Union[str, int] = None, count=None):
...
+            sep_columns: charsequence which separates columns
+            suggestions: self-explanatory
```
Comment (rcorre) [inline qutebrowser/browser/commands.py]: This definitely wasn't self-explanatory to me at first glance, it took me a while to figure out what all these args mean (with the assistance of reading the code).

File: qutebrowser/browser/commands.py (hunk 5 above)
Comment (rcorre) [reply qutebrowser/browser/commands.py]: Actually I guess completion functions can't parse flags yet, but I think that would be the way to go once that is implemented

File: qutebrowser/browser/commands.py [hunk 7]
```diff
@@ -1175,6 +1175,25 @@ def tab_move(self, index: typing.Union[str, int] = None, count=None):
         cmdutils.check_overflow(new_idx, 'int')
         self._tabbed_browser.tabBar().moveTab(cur_idx, new_idx)
 
+    @cmdutils.register(instance='command-dispatcher', scope='window')
```
Comment (rcorre) [inline qutebrowser/browser/commands.py]: Should probably have `hidden=True` as you woudln't call it manually

File: qutebrowser/browser/commands.py (hunk 1 above)
Comment (The-Compiler) [reply qutebrowser/browser/commands.py]: Not the same thing though - that's qutebrowser's window ID (which is just a counter starting with 0), while here, we want the native OS window ID (to e.g. make mpv draw over it).

Comment (rcorre) [reply qutebrowser/browser/commands.py]: Ah, got it

File: doc/help/settings.asciidoc [hunk 8]
```diff
@@ -223,6 +223,7 @@
 |<<scrolling.bar,scrolling.bar>>|Show a scrollbar.
 |<<scrolling.smooth,scrolling.smooth>>|Enable smooth scrolling for web pages.
 |<<session.default_name,session.default_name>>|Name of the session to save by default.
+|<<session.lazy_restore,session.lazy_restore>>|Load a restored tab as soon as it takes focus.
```
Comment (The-Compiler) [inline doc/help/settings.asciidoc]: This is an unrelated change, and it's already like this in `master` - did you base your PR on some older `master` branch or something?

File: misc/userscripts/iframe [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+    __getstaticattr__ is the static equivalent of __getattr__.
+    """
+    def __getattr__(cls, name):
+        return cls.__getstaticattr__(name)
+
+
+class Env(metaclass=StaticGetAttrType):
+    """Helper class which allows to access environment variables"""
+    def __getstaticattr__(name):
+        return os.environ[name]
```
Comment (The-Compiler) [inline misc/userscripts/iframe]: This seems to be a lot of magic just to avoid writing `os.environ` below. Really, just use `os.environ['QUTE_FIFO']` etc. instead :wink:

File: misc/userscripts/iframe [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+
+
+def read(path):
+    """Reads every line of a file
+
+    Return:
+        content or None
```
Comment (The-Compiler) [inline misc/userscripts/iframe]: When would this return `None`?

File: misc/userscripts/iframe [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+    """
+    with open(path) as f:
+        return '\n'.join(f.readlines())
+
+
+def qexec(*cmds):
```
Comment (The-Compiler) [inline misc/userscripts/iframe]: You might want to consider [python-qutescript](https://github.com/hiway/python-qutescript) instead, which also allows for nicer access to the environment variables :wink:

File: qutebrowser/browser/commands.py [hunk 12]
```diff
@@ -1175,6 +1175,25 @@ def tab_move(self, index: typing.Union[str, int] = None, count=None):
...
+    @cmdutils.argument('choice', completion=miscmodels.suggest)
+    def suggest(self, command: str, columns: int,
+                sep_suggestions: str, sep_columns: str,
+                suggestions: str, choice: str):
+        """Allows userscripts to let a user choose between extracted information.
```
Comment (The-Compiler) [inline qutebrowser/browser/commands.py]: "extracted information"?

File: qutebrowser/browser/commands.py (hunk 1 above)
Comment (The-Compiler) [reply qutebrowser/browser/commands.py]: I wonder if this should be called `QUTE_NATIVE_WID` or so to make it clearer.

File: qutebrowser/browser/commands.py [hunk 13]
```diff
@@ -1231,6 +1250,12 @@ def _run_userscript(self, cmd, *args, verbose=False):
             'QUTE_MODE': 'command',
         }
 
+        try:
+            # not sure if int(winId()) is safe to call on every os.
+            env['QUTE_WID'] = str(int(self._tabbed_browser.winId()))
+        except Exception:
+            pass
```
Comment (The-Compiler) [inline qutebrowser/browser/commands.py]: Please don't do this. If you actually see an exception happening, catch that exception - if not, just add it to `env` above, and let the tests take care of ensuring it works.

File: qutebrowser/commands/cmdutils.py [hunk 14]
```diff
@@ -31,6 +31,22 @@
 cmd_dict = {}
 
 
+def quote(s):
```
Comment (The-Compiler) [inline qutebrowser/commands/cmdutils.py]: You'll need to adjust failing tests in `tests/unit/completion/test_completer.py` for this move. Also, can you please add some tests for this to `tests/unit/commands/test_cmdutils.py` as `cmdutils.py` is one of the files where tests are enforced?

File: qutebrowser/misc/sessions.py [hunk 15]
```diff
@@ -422,7 +422,7 @@ def load(self, name, temp=False):
                                         window=window.win_id)
             tab_to_focus = None
             for i, tab in enumerate(win['tabs']):
-                new_tab = tabbed_browser.tabopen()
+                new_tab = tabbed_browser.tabopen(background=False)
```
Comment (The-Compiler) [inline qutebrowser/misc/sessions.py]: This seems like another unrelated change I've done in `master`.

File: qutebrowser/completion/models/miscmodels.py [hunk 16]
```diff
@@ -47,6 +47,31 @@ def helptopic(*, info):
     return model
 
 
+def suggest(*args, **kwargs):
```
Comment (rcorre) [inline qutebrowser/completion/models/miscmodels.py]: Don't forget to add a test in `test_models.py`

**Comment by seebye** [comment]
> Should probably have hidden=True as you woudln't call it manually
@rcorre I don't get where to add it, adding it to
```python
@cmdutils.register(instance='command-dispatcher', scope='window')
```
leads to a crash on my device.

File: qutebrowser/browser/commands.py (hunk 7 above)
Comment (seebye) [reply qutebrowser/browser/commands.py]: I don't get where to add it, adding it to
```python
@cmdutils.register(instance='command-dispatcher', scope='window')
```
leads to a crash on my device.

File: doc/help/settings.asciidoc (hunk 8 above)
Comment (seebye) [reply doc/help/settings.asciidoc]: Right, I didn't update the code since my last pull request.

File: qutebrowser/browser/commands.py (hunk 7 above)
Comment (rcorre) [reply qutebrowser/browser/commands.py]: Sorry, `hide=True`

Comment (seebye) [reply qutebrowser/browser/commands.py]: Also leads to a crash.
Decorator saves kwargs: https://github.com/qutebrowser/qutebrowser/blob/master/qutebrowser/commands/cmdutils.py#L77
Command creation: https://github.com/qutebrowser/qutebrowser/blob/master/qutebrowser/commands/cmdutils.py#L111
Command constructor doesn't allow to assign an argument to a parameter named hide/hidden: https://github.com/qutebrowser/qutebrowser/blob/b31360b6e3cdd1bda27f63c6dcc42934ce27db70/qutebrowser/commands/command.py#L78
The only way to hide a command seems to be debug=True (or to exclude it to be used in a mode).

Comment (rcorre) [reply qutebrowser/browser/commands.py]: Argh, you're right. We just hide commands that aren't valid in normal mode. Ah well, not a big deal.

File: qutebrowser/browser/commands.py (hunk 1 above)
Comment (seebye) [reply qutebrowser/browser/commands.py]: I think it's also interesting to mention that qutebrowser creates a window per tab, so third party programs could simply draw above a tab and therefore a user could switch between e.g. watch a video and doing something else in another tab.
The window id of the currently focused tab is stored in:
```python
int(self._tabbed_browser._now_focused.winId())
```

File: qutebrowser/browser/commands.py [hunk 17]
```diff
@@ -1175,6 +1175,35 @@ def tab_move(self, index: typing.Union[str, int] = None, count=None):
         cmdutils.check_overflow(new_idx, 'int')
         self._tabbed_browser.tabBar().moveTab(cur_idx, new_idx)
 
+    @cmdutils.register(instance='command-dispatcher', scope='window',
+                       debug=True)
```
Comment (The-Compiler) [inline qutebrowser/browser/commands.py]: I don't think the `debug=True` makes sense here. I guess it'll just show up in the completion, no biggie.

File: qutebrowser/browser/commands.py (hunk 7 above)
Comment (The-Compiler) [reply qutebrowser/browser/commands.py]: There used to be a `hide=True`, but it confused a lot of people, so I removed it.

**Comment by The-Compiler** [comment]
Note that I'm currently busy with exams until early February - due to that, it might take a bit longer (probably until then) before I can get to reviewing this again.

**Comment by The-Compiler** [comment]
I planned to take a look at this during my holidays before starting to learn for exams, but I didn't manage to do so (mainly because of a lot of issues with Qt 5.11...). I'll get back to it in September though!

Sorry for the loooong delay - there's just a lot of stuff piling up. FWIW I'm not too happy about how the `:suggest` command is implemented, though - it's a lot of manual parsing via regexes which is quite hard to understand/maintain. I really wonder whether it's a good idea to add this at all, or if we should rather wait for a proper plugin API for things like this...

**Comment by seebye** [comment]
> I planned to take a look at this during my holidays before starting to learn for exams, but I didn't manage to do so (mainly because of a lot of issues with Qt 5.11...).
No problem, but I'll soon have exams myself.
So I can't work on it currently.

>  it's a lot of manual parsing via regexes which is quite hard to understand/maintain.
I'm also not happy with the implementation as far as I remember I planed to implement it via str.split(r'(?!\\)(\\\\)*;') which is not hard to understand.  
However unfortunately the split command is not able to do that as explained in the comments.  
(So it mutated to the current implementation.)

> I really wonder whether it's a good idea to add this at all, or if we should rather wait for a proper plugin API for things like this...
I'm already using this by injecting code with the config.py, so I'm also fine with not merging it.

**Comment by The-Compiler** [comment]
After thinking some more about it, I'm going to close this - sorry! I think something like the `:suggest` command is a great idea for a future plugin API though, so I added it to #30.
